[Enhancement] Abstract rating block

octaedro opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR refactors the `rating` and `rating stars` blocks.

The blocks [Rating](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/rating/block.tsx) and [Rating stars](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/rating-stars/block.tsx) share a lot of code, so this PR abstracts part of that logic to the [component ProductRating](https://github.com/woocommerce/woocommerce/pull/50810/files#diff-ac3120cf2d2462de53cd9740970e5567db4a0c65b839c79d638370dba98b8040R138).

The blocks `All Products` and `Single Product` use the block `rating`.

Closes #42582.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a few products.
2. Go to the product pages and review them (including a rating). 
3. Add a new page or post.
4. Add an `All Products` block and verify that the products with ratings look good.

https://github.com/user-attachments/assets/fcec1e86-9681-4545-82cb-daf9f1168284

5. Remove that block and add a `Single Product` block.
6. Verify the rating looks correct.
7. Click the rating and press `Enter`.
8. Paste the following shortcode
```
<!-- wp:woocommerce/product-rating-stars /-->
```
9. Press the three dots and `Edit as HTML`.
10. Paste the shortcode again and press `Edit visually`.
11. Press the three dots next to the `Attempt Block Recovery` button and select `Resolve`.
12. Press `Convert to Blocks`.

https://github.com/user-attachments/assets/900b3de9-a78c-4090-8a33-690955e4b672

13. That will add the `Rating Stars` block.
14. Verify that `Rating Stars` block looks correct as well.


<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

octaedro committed:
Create rating component

octaedro committed:
Refactor rating block

octaedro committed:
Fix ProductRating component

octaedro committed:
Refactor rating blocks

octaedro committed:
Add changelog

octaedro committed:
Review count maybe visible for rating stars

octaedro committed:
Rename props

thealexandrelara reviewed (changes requested):
Thank you for working on this @octaedro! Code looks good, but I did find some things that might require some fixes.

## Test summary:

✅ Scenario 1: Merchant creates a new post and adds the All Products block. The store contains products with reviews.
- Expected Outcome: The products with ratings look good within the All Products block.
- Details: The All Products block correctly renders the Product Rating block for products with reviews.

---


✅ Scenario 2: Merchant creates a new post and adds the Single Product block. They select a product with reviews.
- Expected Outcome: The Product Rating block correctly shows the rating and the amount of reviews for that product.
- Details: The rating and amount of reviews for the select product are shown correctly.

---

⚠️ Scenario 3: Merchant visits the Single Product template on the editor.
- Expected Outcome: The Product Rating block displays "No Reviews".
- Details: Currently, when the merchant visits the Single Product template in the Editor, the Product Ratings block adds a 0 (zero) to the end of the "No Reviews" text.

![CleanShot 2024-08-30 at 12 46 00@2x](https://github.com/user-attachments/assets/86b7df9b-166c-4d04-9e90-dae35a0af086)

---

⚠️ Scenario 4: Merchant edit/creates a new post or page and adds a Single Product block to it.
- Expected Outcome: The Product Rating block within the Single Product block displays "No Reviews" for products with no reviews.
- Details: Currently, when the merchant edit/creates a new post or page and adds a Single Product block to it, the Product Ratings block adds a 0 (zero) to the end of the "No Reviews" text.

![CleanShot 2024-08-30 at 12 55 02@2x](https://github.com/user-attachments/assets/f71d6985-fff6-4c9f-854d-f28c100bb1ea)

---

⚠️ Scenario 5: Merchant uses a shortcode (`<!-- wp:woocommerce/product-rating-stars /-->`) to add the Product Rating Stars block within the Single Product block.
- Expected Outcome: The Product Rating Stars block must display with the correct styles and only display the stars.
- Details: The amount of reviews is not correctly aligned with the stars, and there is no space between them. The amount of reviews is being displayed despite that I believe it shouldn't, right?
- Comments: By default, the Single Product block uses the Product Rating block. In your testing instructions, you requested adding the Product Rating Stars within the Single Product block. I’d like to know if this is the expected behavior for this block. In this case, should we set the Product Rating Stars as the default block for the Single Product block?

https://github.com/user-attachments/assets/07423b1d-9858-4458-b962-1e527c52b9a4



---

octaedro committed:
Remove 0 after No reviews message

octaedro committed:
Remove review count from rating-stars

octaedro commented:
Thank you @thealexandrelara for your review,

I just fixed the scenarios you mentioned.

> Details: The amount of reviews is not correctly aligned with the stars, and there is no space between them. The amount of reviews is being displayed despite that I believe it shouldn't, right?

That's correct. The Rating stars don't display the review count. Fixed in the commit: [8dcf4dc](https://github.com/woocommerce/woocommerce/pull/50810/commits/8dcf4dc5ad6081e0139a62e9cb212850610454dd)

> Comments: By default, the Single Product block uses the Product Rating block. In your testing instructions, you requested adding the Product Rating Stars within the Single Product block. I’d like to know if this is the expected behavior for this block. In this case, should we set the Product Rating Stars as the default block for the Single Product block?

I would continue using the same rating block we were using. I wanted to add the stars there just make testing easier.

Could you take another look at this PR?

thealexandrelara reviewed (approved):
Thank you for fixing the raised issues. Everything LGTM!  Great work! 🚀 

## Test summary:

✅ Scenario 1: Merchant creates a new post and adds the All Products block. The store contains products with reviews.
- Expected Outcome: The products with ratings look good within the All Products block.
- Details: The All Products block correctly renders the Product Rating block for products with reviews.

---

✅ Scenario 2: Merchant creates a new post and adds the Single Product block. They select a product with reviews.
- Expected Outcome: The Product Rating block correctly shows the rating and the amount of reviews for that product.
- Details: The rating and amount of reviews for the select product are shown correctly.

---

✅  Scenario 3: Merchant visits the Single Product template on the editor.
- Expected Outcome: The Product Rating block displays "No Reviews".
- Details: When the merchant visits the Single Product template in the Editor, the Product Ratings block correctly displays the "No Reviews" text.

---

✅  Scenario 4: Merchant edit/creates a new post or page and adds a Single Product block to it.
- Expected Outcome: The Product Rating block within the Single Product block displays "No Reviews" for products with no reviews.
- Details: When the merchant visits the Single Product template in the Editor, the Product Ratings block correctly displays the "No Reviews" text.

---

✅  Scenario 5: Merchant uses a shortcode (`<!-- wp:woocommerce/product-rating-stars /-->`) to add the Product Rating Stars block within the Single Product block.
- Expected Outcome: The Product Rating Stars block must display with the correct styles and only display the stars.
- Details: The block correctly displays only the stars when the product contains reviews. If the product does not contain any review, then it shows the "No Reviews" text correctly aligned to an empty star icon.

---
